framework/src/utils/SlepcSupport.C

```diff
@@ -549,73 +550,145 @@ mooseEPSFormMatrices(EigenProblem & eigen_problem, EPS eps, Vec x, void * ctx)
-
-  PetscVector<Number> X_global(x, sys.comm());
+  auto & nl = eigen_problem->getCurrentNonlinearEigenSystem();
+  auto & sys = nl.sys();
+  auto & dof_map = sys.get_dof_map();
+
+#ifndef NDEBUG
+  auto & petsc_mat = cast_ref<PetscMatrix<Number> &>(libmesh_mat);
+  mooseAssert(!dof_map.n_constrained_dofs() == (mat == petsc_mat.mat()),
+              "If we do not have constrained dofs, then mat and libmesh_mat should be the same. "
+              "Conversely, if we do have constrained dofs, they must be different");
+#endif
 
-  PetscVector<Number> & X_sys = *cast_ptr<PetscVector<Number> *>(sys.solution.get());
+  updateCurrentLocalSolution(sys, x);
 
-  // Use the system's update() to get a good local version of the
-  // parallel solution.  This operation does not modify the incoming
-  // "x" vector, it only localizes information from "x" into
-  // sys.current_local_solution.
-  X_global.swap(X_sys);
-  sys.update();
-  X_global.swap(X_sys);
+  if (!eigen_problem->constJacobian())
+    libmesh_mat.zero();
 
-  std::vector<std::unique_ptr<SparseMatrix<Number>>> jacobians;
+  eigen_problem->computeJacobianTag(*sys.current_local_solution.get(), libmesh_mat, tag);
 
-  for (auto & mat : mats)
+  if (dof_map.n_constrained_dofs())
   {
-    jacobians.emplace_back(std::make_unique<PetscMatrix<Number>>(mat, sys.comm()));
-    jacobians.back()->attach_dof_map(sys.get_dof_map());
-    if (!eigen_problem->constJacobian())
-      jacobians.back()->zero();
+    PetscMatrix<Number> sub(mat, sys.comm());
+    sys.copy_super_to_sub(libmesh_mat, sub);
   }
+}
 
-  eigen_problem->computeMatricesTags(*sys.current_local_solution.get(), jacobians, tags);
+void
+moosePetscSNESFormMatricesTags(SNES /*snes*/,
+                               Vec x,
+                               std::vector<Mat> & mats,
+                               std::vector<SparseMatrix<Number> *> & libmesh_mats,
+                               void * ctx,
+                               const std::set<TagID> & tags)
+{
+  EigenProblem * eigen_problem = static_cast<EigenProblem *>(ctx);
+  auto & nl = eigen_problem->getCurrentNonlinearEigenSystem();
+  auto & sys = nl.sys();
+  auto & dof_map = sys.get_dof_map();
+
+#ifndef NDEBUG
+  for (const auto i : index_range(mats))
+    mooseAssert(!dof_map.n_constrained_dofs() ==
+                    (mats[i] == cast_ptr<PetscMatrix<Number> *>(libmesh_mats[i])->mat()),
+                "If we do not have constrained dofs, then mat and libmesh_mat should be the same. "
+                "Conversely, if we do have constrained dofs, they must be different");
+#endif
+
+  updateCurrentLocalSolution(sys, x);
+
+  for (auto * const lm_mat : libmesh_mats)
+    if (!eigen_problem->constJacobian())
+      lm_mat->zero();
+
+  eigen_problem->computeMatricesTags(*sys.current_local_solution.get(), libmesh_mats, tags);
+
+  if (dof_map.n_constrained_dofs())
+    for (const auto i : index_range(mats))
+    {
+      PetscMatrix<Number> sub(mats[i], sys.comm());
```

Comment:
Where is this sub used?

Reply:
the sub-matrices (the matrices representing only unconstrained dofs) are the matrices used in the eigensolves in SLEPc